[LC-609] Add unit tests (Transactions)

metalg0su opened the pull request:
# Goal
Deep dive into transactions test.

## Changes
- Increase coverage: 87% -> 95%+
- Add test cases: 30 -> 160+
- Change `$ make unit-test` behavior to show skip and xfail reasons
> Test Summary Notes:
> - . : Success
> - x: Expected failure and failed.
> - X: Expected failure but succeeded.
> - s: Skipped

metalg0su committed:
[LC-609] Add tests for TransactionVersioner

metalg0su committed:
[LC-609] Add tests for TransactionSerializer

metalg0su committed:
[LC-609] Add tests for TransactionVerifier

metalg0su committed:
[LC-609] Add tests for TransactionBuilder

metalg0su committed:
[LC-609] Add tests for Transaction

metalg0su committed:
[LC-609] Integrate original unit tests

- Move tx_verifier_cache test and refactoring it

metalg0su committed:
[LC-609] Add fixtures for transactions

metalg0su committed:
[LC-609] Add dependencies for tests

- Change test command: show xfail reason
> This option reveals various execuses which justify broken tests.

metalg0su commented on testcase/unittest/blockchain/conftest.py:
```diff
@@ -0,0 +1,102 @@
+        # Attributes to be generated
+        tx_builder.from_address = ExternalAddress.fromhex_address(tx_builder.signer.address)
+        tx_builder.hash = Hash32(os.urandom(Hash32.size))
+        # tx_builder.signature = Signature(Signature.size)
```
I'm not sure that the signature is needed or not.

metalg0su commented on testcase/unittest/blockchain/transactions/test_transaction_serializer.py:
```diff
@@ -0,0 +1,280 @@
+class TestTransactionSerializer_genesis:
+    tx_version = genesis.version
+
+    # TODO: What is diffrence between tx.raw_data and its dict casted?
```
What is diffrence between tx.raw_data and its dict casted?

zsaladin replied:
It is dangerous to return `raw_data` without copying.
Assume that `raw_data` is manipulated out of transaction. Then the transaction will lose integrity.

metalg0su replied:
I see.

metalg0su replied:
-> fixed by casting raw_data to dict:
https://github.com/icon-project/loopchain/blob/92b74a99952fa92f12caa81fad8f648f4ca55524/loopchain/blockchain/transactions/genesis/transaction_serializer.py#L19

zsaladin commented on testcase/unittest/blockchain/transactions/test_transaction_builder.py:
```diff
@@ -0,0 +1,181 @@
+        assert expected_nid == nid_generated
+
+    @pytest.mark.parametrize("tx_hash, expected_nid", [
+        (genesis.NTxHash.mainnet, genesis.NID.mainnet),
```
These values are written here inline.
Because the tests will pass even if these values are changed from original codes.

zsaladin commented on testcase/unittest/blockchain/transactions/test_transaction_verifier_cache.py:
```diff
@@ -16,29 +15,16 @@
+    def test_no_cache_attr_in_tx_if_not_verified(self, tx_version, tx_factory, target_attr):
         """Check that Transaction has no attribute until the verification func has been invoked"""
+        tx = tx_factory(tx_version=tx_version)
         if "cache" in dir(tx):
```
```python
if any("cache" in attr for attr in dir(tx):
```

metalg0su replied:
I was definitely wrong. This test catches only `Transaction.cache` attribute.

metalg0su commented on testcase/unittest/blockchain/conftest.py:
```diff
@@ -0,0 +1,86 @@
+                "- Mille Plateaux, Gilles Deleuze & Felix Guattari\n\n\"" \
+                "Hyperconnect the world\""
+
+        if tx_version == v2.version:
```
Oh, You're right. I made them to be evaluate all versions!!! Thanks

metalg0su commented on testcase/unittest/blockchain/conftest.py:
```diff
@@ -0,0 +1,86 @@
+            tx_builder.fee: int = int(0.01 * 10**18)
+            tx_builder.nonce: int = 10000
+
+        if tx_version == v3.version:
```
Thank you.

metalg0su committed:
[LC-609] Apply reviews

 - Replace multiple if evaluations to elif
 - Remove redundant inline if
 - Fix test in tx_cache
 - Add type hints at TC signature
 - Change file names (original test -> test_transactions.py, new
functional test -> test_transaction.py)
 - ... and Refactorings

Co-authored-by: Daehee Kim <zsaladinz@gmail.com>
Co-authored-by: Seungmin Ryu <yakkle@gmail.com>

metalg0su commented on testcase/unittest/blockchain/transactions/test_transaction_builder.py:
```diff
@@ -0,0 +1,181 @@
+
+    def test_reset_cache_resets_members(self, tx_builder_factory, tx_version):
+        tx_builder = tx_builder_factory(tx_version=tx_version)
+        tx_builder.from_address = tx_builder.from_address if tx_builder.from_address else "aa"
```
How silly am I --;

metalg0su committed:
[LC-609] Fix raw_data conversion in genesis serializer

metalg0su commented on testcase/unittest/blockchain/conftest.py:
```diff
@@ -0,0 +1,86 @@
+
+    pytest.tx_versioner = TransactionVersioner()
+
+
```
Agree!

metalg0su commented on testcase/unittest/blockchain/transactions/test_transaction_.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
```
Changed original test as `test_transaction.py` and new functional test as `test_transaction.py`

yakkle commented on testcase/unittest/blockchain/conftest.py:
```diff
@@ -0,0 +1,93 @@
+            "Where are you going? Where are you coming from? What are you heading for? " \
+            "These are totally useless questions.\n\n " \
+            "- Mille Plateaux, Gilles Deleuze & Felix Guattari\n\n\"" \
+            "Hyperconnect the world\""
```
my suggestion is,
```py
        tx_builder.message = (
            "A rhizome has no beginning or end; "
            "it is always in the middle, between things, interbeing, intermezzo. "
            "The tree is filiation, but the rhizome is alliance, uniquely alliance. "
            "The tree imposes the verb \"to be\" but the fabric of the rhizome is the conjunction, \"and ... and ...and...\""
            "This conjunction carries enough force to shake and uproot the verb \"to be.\" "
            "Where are you going? Where are you coming from? What are you heading for? "
            "These are totally useless questions.\n\n "
            "- Mille Plateaux, Gilles Deleuze & Felix Guattari\n\n\""
            "Hyperconnect the world\"")
```

metalg0su replied:
your suggestion is, good.
